src/common/substrait/src/df_expr.rs

```diff
@@ -0,0 +1,728 @@
+                asc: false,
+                nulls_first: false,
+            }
+        }
+        "sort_asc" => {
+            ensure_arg_len(1)?;
+            Expr::Sort {
+                expr: Box::new(inputs.pop_front().unwrap()),
+                asc: true,
+                nulls_first: false,
+            }
+        }
+        // those are datafusion built-in "scalar functions".
+        "abs"
+        | "acos"
+        | "asin"
+        | "atan"
+        | "atan2"
+        | "ceil"
+        | "cos"
+        | "exp"
+        | "floor"
+        | "ln"
+        | "log"
+        | "log10"
+        | "log2"
+        | "power"
+        | "pow"
+        | "round"
+        | "signum"
+        | "sin"
+        | "sqrt"
+        | "tan"
+        | "trunc"
+        | "coalesce"
+        | "make_array"
+        | "ascii"
+        | "bit_length"
+        | "btrim"
+        | "char_length"
+        | "character_length"
+        | "concat"
+        | "concat_ws"
+        | "chr"
+        | "current_date"
+        | "current_time"
+        | "date_part"
+        | "datepart"
+        | "date_trunc"
+        | "datetrunc"
+        | "date_bin"
+        | "initcap"
+        | "left"
+        | "length"
+        | "lower"
+        | "lpad"
+        | "ltrim"
+        | "md5"
+        | "nullif"
+        | "octet_length"
+        | "random"
+        | "regexp_replace"
+        | "repeat"
+        | "replace"
+        | "reverse"
+        | "right"
+        | "rpad"
+        | "rtrim"
+        | "sha224"
+        | "sha256"
+        | "sha384"
+        | "sha512"
+        | "digest"
+        | "split_part"
+        | "starts_with"
+        | "strpos"
+        | "substr"
+        | "to_hex"
+        | "to_timestamp"
+        | "to_timestamp_millis"
+        | "to_timestamp_micros"
+        | "to_timestamp_seconds"
+        | "now"
+        | "translate"
+        | "trim"
+        | "upper"
+        | "uuid"
+        | "regexp_match"
+        | "struct"
+        | "from_unixtime"
+        | "arrow_typeof" => Expr::ScalarFunction {
+            fun: BuiltinScalarFunction::from_str(fn_name).unwrap(),
+            args: inputs.into(),
+        },
+        // skip ScalarUDF, unimplemented.
+        // skip AggregateFunction, is covered in substrait::AggregateRel
+        // skip WindowFunction, is covered in substrait WindowFunction
+        // skip AggregateUDF, unimplemented.
+        // skip InList, unimplemented
+        // skip Wildcard, unimplemented.
+        // end other direct expr
+        _ => UnsupportedExprSnafu {
+            name: format!("scalar function {}", fn_name),
+        }
+        .fail()?,
+    };
+
+    Ok(expr)
+}
+
+/// Convert DataFusion's `Expr` to substrait's `Expression`
+pub fn expression_from_df_expr(
+    ctx: &mut ConvertorContext,
+    expr: &Expr,
+    schema: &Schema,
+) -> Result<Expression> {
+    let expression = match expr {
+        // Don't merge them with other unsupported expr arms to preserve the ordering.
```

Comment:
Preserves the ordering the same with DataFusion `Expr` definition. This is to make it easy to maintain because this enumuration is likely to change in the future.